test/Dialect/FIRRTL/lower-layers.mlir

```diff
@@ -1,11 +1,353 @@
+   %foo_o = firrtl.instance foo @ColoredPorts(out o: !firrtl.probe<uint<1>, @A>)
+  }
+
+  // CHECK-LABEL: @ColoredThings
+  firrtl.module @ColoredThings() {
+    // CHECK: %0 = firrtl.wire : !firrtl.probe<bundle<f: uint<1>>>
+    %0 = firrtl.wire : !firrtl.probe<bundle<f: uint<1>>, @A>
+    // CHECK: %1 = firrtl.ref.sub %0[0] : !firrtl.probe<bundle<f: uint<1>>>
+    %1 = firrtl.ref.sub %0[0] : !firrtl.probe<bundle<f: uint<1>>, @A>
+    // CHECK-NOT: firrtl.cast
+    %2 = firrtl.ref.cast %1 : (!firrtl.probe<uint<1>, @A>) -> !firrtl.probe<uint<1>, @A::@B>
+  }
+
+  //===--------------------------------------------------------------------===//
+  // Removal of Enabled Layers
+  //===--------------------------------------------------------------------===//
+
+  // CHECK-LABEL: @EnabledLayers() {
+  firrtl.module @EnabledLayers() attributes {layers = [@A]} {}
+
+  // CHECK-LABEL: @EnabledLayersOnInstance()
+  firrtl.module @EnabledLayersOnInstance() attributes {layers = [@A]} {
+    // CHECK: firrtl.instance enabledLayers @EnabledLayers()
+    firrtl.instance enabledLayers {layers = [@A]} @EnabledLayers()
+  }
+
+  //===--------------------------------------------------------------------===//
+  // Removal of Layerblocks and Layers
+  //===--------------------------------------------------------------------===//
+
+  // CHECK-NOT: firrtl.layer @GoodbyeCruelWorld
+  firrtl.layer @GoodbyeCruelWorld bind {}
+
+  // CHECK-LABEL @WithLayerBlock
+  firrtl.module @WithLayerBlock() {
+    // CHECK-NOT firrtl.layerblock @GoodbyeCruelWorld
+    firrtl.layerblock @GoodbyeCruelWorld {
+    }
+  }
+
+  //===--------------------------------------------------------------------===//
+  // Capture
+  //===--------------------------------------------------------------------===//
+
+  // CHECK: firrtl.module private @CaptureHardware_A(in %_: !firrtl.uint<1>, in %__0: !firrtl.uint<1>) attributes {portNames = ["_", "_"]} {
+  // CHECK:   %0 = firrtl.add %_, %__0 : (!firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<2>
+  // CHECK: }
+  // CHECK: firrtl.module @CaptureHardware() {
+  // CHECK:   %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
+  // CHECK:   %c0_ui1_0 = firrtl.constant 0 : !firrtl.uint<1>
+  // CHECK:   %captureHardware_A__, %captureHardware_A___1 = firrtl.instance captureHardware_A {lowerToBind, output_file = #hw.output_file<"groups_Test_A.sv", excludeFromFileList>} @CaptureHardware_A(in _: !firrtl.uint<1>, in _: !firrtl.uint<1>)
+  // CHECK:   firrtl.strictconnect %captureHardware_A___1, %c0_ui1_0 : !firrtl.uint<1>
+  // CHECK:   firrtl.strictconnect %captureHardware_A__, %c0_ui1 : !firrtl.uint<1>
+  // CHECK: }
+  firrtl.module @CaptureHardware() {
+    %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
+    %c1_ui1 = firrtl.constant 0 : !firrtl.uint<1>
+    firrtl.layerblock @A {
+      %0 = firrtl.add %c0_ui1, %c1_ui1 : (!firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<2>
+    }
+  }
```

Comment:
General comment throughout the tests: the names of ports are intentionally being generated by the pass as anonymous. While I can see some benefit in checking the names if the names are intentionally generated to be "good", explicitly checking things like `%captureHardware_A__` and `%captureHardware_A__1` is likely just going to make the test hard to maintain.

Other test nits: 

1. Checking things like the `portNames` attribute array is similarly tedious to maintain.
2. I will usually skip checking the types unless there is a good reason to.
3. If the constants don't change, then I would advise skipping including them in the checks.
4. If something should follow immediately after something else, then a `CHECK-NEXT` is always better than a `CHECK`.

Question about this test: why is the constant changing? This seems to be changing a c1 to a c0?

Reply:
> Question about this test: why is the constant changing? This seems to be changing a c1 to a c0?

The test is just using a non-canonical form for a constant zero. I think this wants to have the value of the second constant be one.

Reply:
I can fix the test so it round-trips nicer :+1:

Reply:
I updated the test to pattern match against the generated names of ports, instances, and modules. I also stopped matching against the port-types in the instance ops, and the port names attributes.

 In the case of these constants, I like having them CHECK'd because I think it makes the expected output more legible.

Reply:
> In the case of these constants, I like having them CHECK'd because I think it makes the expected output more legible.

The PR _could_ get some of that back by using indentation of the checked strings. Consider some of the techniques used in this test to keep it legible: https://github.com/llvm/circt/blob/main/test/Dialect/FIRRTL/grand-central.mlir